ENH: Make roc_curve array API compatible

lithomas1 opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

xref #26024 

#### What does this implement/fix? Explain your changes.

Makes roc_curve array API compatible.

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
https://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

lithomas1 committed:
ENH: Make roc_curve array API compatible

lithomas1 committed:
add whatsnew

lithomas1 committed:
Merge branch 'main' of github.com:scikit-learn/scikit-learn into roc-curve

lithomas1 committed:
try to fix for oldest numpy

lithomas1 committed:
fix argsort for old numpy

OmarManzoor commented on sklearn/utils/_array_api.py:
```diff
@@ -452,6 +458,34 @@ def isdtype(self, dtype, kind):
         return numpy.power(x1, x2)
 
+    def cumulative_sum(self, x, axis=None, dtype=None, include_initial=False):
+        # This is copy pasted from array-api-compat
```
Is this required to support older versions of numpy?

lithomas1 replied:
Yeah, cumulative_sum is new in numpy 2.1.

OmarManzoor replied:
I am not sure about that. Maybe @ogrisel or @betatim would have a better idea.

lucyleeow replied:
Is `cumulative_sum` not in array api? https://data-apis.org/array-api/latest/API_specification/generated/array_api.cumulative_sum.html#array_api.cumulative_sum

lithomas1 replied:
It is. 
The issue happens though array API is turned off (and array-api-compat is not installed), since we need to rely on what's in numpy/what we have wrapped in the ``_NumpyAPIWrapper`` then.

(I could be misunderstanding how NumpyAPIWrappers is supposed to function though.)

lucyleeow replied:
Ah good point, they should add a version added note in their docs!

I actually also have this problem over in #29431

OmarManzoor replied:
I think the PR for vendoring the array-api-compat and array-api-extra is close to being merged. So let's wait for that.

OmarManzoor commented on sklearn/utils/validation.py:
```diff
@@ -2165,17 +2165,19 @@ def _check_sample_weight(
-    if dtype is not None and dtype not in [np.float32, np.float64]:
-        dtype = np.float64
+    if dtype is not None and dtype not in [xp.float32, xp.float64]:
+        dtype = xp.float64
```
This also might need to be adjusted using the maximum floating dtype helper

OmarManzoor reviewed (commented):
Thanks for the PR @lithomas1 

I added some initial comments.

lithomas1 committed:
update following code review

lithomas1 committed:
fix mps

lithomas1 commented:
Thanks for the review, and sorry for the slow reply.

I addressed the device issues (a MPS run on my Intel MBP uncovered some more issues that I fixed).

OmarManzoor commented on sklearn/metrics/_ranking.py:
```diff
@@ -838,29 +845,35 @@ def _binary_clf_curve(y_true, y_score, pos_label=None, sample_weight=None):
-    distinct_value_indices = np.where(np.diff(y_score))[0]
-    threshold_idxs = np.r_[distinct_value_indices, y_true.size - 1]
+    distinct_value_indices = xp.nonzero(y_score[1:] - y_score[:-1])[0]
+    # can't use size since PyTorch size is a method and not a property
+    # as specified by the array API spec
```
I think we can remove the comment
```suggestion
```

OmarManzoor commented on sklearn/metrics/_ranking.py:
```diff
@@ -1160,27 +1174,34 @@ def roc_curve(
     # get dtype of `y_score` even if it is an array-like
-    thresholds = np.r_[np.inf, thresholds]
+    thresholds = xp.astype(thresholds, _max_precision_float_dtype(xp, device))
+    thresholds = xp.concat([xp.asarray([xp.inf], device=device), thresholds])
```
Do we really need to convert scalars to arrays?

lithomas1 replied:
I think I made this fix as part of a fix for the MPS device.

I'll double check after the array-api-compat/extra PR gets merged in.

lithomas1 replied:
I looked again, and it doesn't look like scalar arguments are supported in array API for concat.

https://data-apis.org/array-api/latest/API_specification/generated/array_api.concat.html#concat

OmarManzoor commented on sklearn/metrics/tests/test_common.py:
```diff
@@ -1867,11 +1867,20 @@ def check_array_api_metric(
+                atol=_atol_for_type(dtype_name),
+            )
+
+        # Handle case where there is multiple return
+        # values, e.g. roc_curve
```
```suggestion
        # Handle cases where there are multiple return
        # values, e.g. roc_curve
```

OmarManzoor reviewed (commented):
Thanks for the updates @lithomas1.
Mostly this looks good. However let's consider waiting for the array-api-extra PR

lithomas1 committed:
Apply suggestions from code review

Co-authored-by: Omar Salman <omar.salman@arbisoft.com>

lithomas1 committed:
Merge branch 'main' into roc-curve

lithomas1 committed:
fix tests

lucascolley commented on sklearn/externals/array_api_extra/_lib/_funcs.py:
```diff
@@ -552,7 +552,7 @@ def isclose(
         )
         if equal_nan:
-            out = xp.where(xp.isnan(a) & xp.isnan(b), xp.asarray(True), out)
+            out = xp.where(xp.isnan(a) & xp.isnan(b), True, out)
```
PR merged, please ping me when you need a release

lucascolley replied:
gh-31080

betatim replied:
Marking this as resolved as part of trying to get an overview where we are. Please unresolve if we need to keep it open

lithomas1 committed:
Merge branch 'main' into roc-curve

OmarManzoor commented on sklearn/utils/extmath.py:
```diff
@@ -1189,10 +1198,31 @@ def stable_cumsum(arr, axis=None, rtol=1e-05, atol=1e-08):
-        out.take(-1, axis=axis), expected, rtol=rtol, atol=atol, equal_nan=True
+    xp, _, device = get_namespace_and_device(arr)
+    if axis is None:
+        # np.cumsum calculates cumsum across flattened array
+        # for axis=None
+        # Let's also flatten to match
+        # (if we don't flatten, cumulative_sum below will raise error)
```
```suggestion
        # np.cumsum calculates cumsum across flattened array
        # for axis=None, which is why we flatten here.
        # (if we don't flatten, cumulative_sum below will raise error)
```

OmarManzoor commented on sklearn/utils/extmath.py:
```diff
@@ -1189,10 +1198,31 @@ def stable_cumsum(arr, axis=None, rtol=1e-05, atol=1e-08):
+    out = xp.cumulative_sum(arr, axis=axis, dtype=max_float_dtype)
+    expected = xp.sum(arr, axis=axis, dtype=max_float_dtype)
+    # workaround to get last element
+    # (negative indices, e.g. -1, behavior is undefined)
```
```suggestion
    # (behavior of negative indices, e.g. -1, is undefined)
```

OmarManzoor reviewed (approved):
LGTM. Thanks @lithomas1 

CC: @ogrisel @betatim for a second review

lithomas1 committed:
Apply suggestions from code review

Co-authored-by: Omar Salman <omar.salman@arbisoft.com>

lithomas1 committed:
remove spurious TODO

lithomas1 committed:
Merge branch 'main' of github.com:scikit-learn/scikit-learn into roc-curve

betatim commented on sklearn/utils/validation.py:
```diff
@@ -2169,17 +2175,18 @@ def _check_sample_weight(
-    if dtype is not None and dtype not in [np.float32, np.float64]:
-        dtype = np.float64
+    if dtype is not None and dtype not in [xp.float32, xp.float64]:
+        dtype = _max_precision_float_dtype(xp, device)
```
Can you update the description of the `dtype` parameter in the docstring? I think we should change the statement that `np.flaot64` will be the dtype of `sample_weight` if nothing else is specified. Maybe "The output will be the highest precision floating point dtype" or some such?

lithomas1 replied:
I reworded this slightly to "Otherwise, the output has the highest precision floating point dtype supported by the array namespace/device of the input arrays."

Let me know if that sounds OK to you.

betatim commented on sklearn/utils/validation.py:
```diff
@@ -2629,13 +2636,14 @@ def _check_pos_label_consistency(pos_label, y_true):
+            or xp.all(classes == xp.asarray([-1], device=device))
+            or xp.all(classes == xp.asarray([1], device=device))
         ):
             classes_repr = ", ".join([repr(c) for c in classes.tolist()])
```
Do you know why `.tolist()`  works here? I don't think it is part of the array API standard and `classes` is a array API array, which is why I am surprised that this works. Are we just getting lucky or does it make sense that this works?

betatim replied:
I'd convert `classes` to a numpy array and then use it here. It should be short, so cheap to convert with `_convert_to_numpy` from `sklearn.utils._array_api`

lithomas1 committed:
fixes from code review

lithomas1 committed:
Merge branch 'main' into roc-curve

OmarManzoor commented:
@lesteve What do you think about merging this PR?

lesteve reviewed (commented):
A few comments

OmarManzoor commented on sklearn/utils/validation.py:
```diff
@@ -2148,9 +2155,10 @@ def _check_sample_weight(
         If None, and `sample_weight` is an array:
 
-            - If `sample_weight.dtype` is one of `{np.float64, np.float32}`,
+            - If `sample_weight.dtype` is one of `{xp.float64, xp.float32}`,
```
I think this point was discussed before and it was decided not to alter the docstrings.

lesteve replied:
OK, just to be explicit @OmarManzoor do you think we should revert these docstring changes then?

OmarManzoor replied:
I think for consistency it would make sense not to modify docstrings unless it is necessary.

OmarManzoor commented on sklearn/utils/extmath.py:
```diff
@@ -1189,10 +1198,29 @@ def stable_cumsum(arr, axis=None, rtol=1e-05, atol=1e-08):
-    expected = np.sum(arr, axis=axis, dtype=np.float64)
-    if not np.allclose(
-        out.take(-1, axis=axis), expected, rtol=rtol, atol=atol, equal_nan=True
+    xp, _, device = get_namespace_and_device(arr)
```
I think no tests were added for this part, since they are mostly util functions that are used elsewhere.

lesteve replied:
In would be more comfortable with some tests. In my experience it's rather easy to write array API code that looks harmless but that is slightly wrong is some edge cases.

betatim replied:
Adding a few tests seems like a sensible idea, especially checking the case of `axis=None` that is now "special"

OmarManzoor commented on sklearn/metrics/_ranking.py:
```diff
@@ -838,29 +845,33 @@ def _binary_clf_curve(y_true, y_score, pos_label=None, sample_weight=None):
-    y_score = y_score[desc_score_indices]
-    y_true = y_true[desc_score_indices]
+    desc_score_indices = xp.argsort(y_score, stable=True, descending=True)
+    y_score = xp.take(y_score, desc_score_indices)
```
I think simple indexing should work in this case as you suggest. But we should try it out to see that it works.

lesteve replied:
OK I pushed the change that I had in mind, the tests pass fine locally, let's see what the CI says about it.

lesteve committed:
Use indexing rather than xp.take

OmarManzoor commented on sklearn/utils/validation.py:
I think you could add some array API code to enhance the current test. Otherwise maybe copy the main part of the test and just compare array API outputs with numpy ones.

betatim commented on sklearn/utils/validation.py:
```diff
@@ -2148,9 +2155,10 @@ def _check_sample_weight(
               then the dtype is preserved.
-            - Else the output has NumPy's default dtype: `np.float64`.
+            - Otherwise, the output has the highest precision floating point dtype
+              supported by the array namespace/device of the input arrays.
```
```suggestion
              supported by the array namespace and device of the input arrays.
```

lithomas1 replied:
I reverted the docstring changes following https://github.com/scikit-learn/scikit-learn/pull/30878/files#r2085052449.

(I assume your OK with it since you thumbs uped the comment, LMK if you still want the suggestion applied).

betatim commented on sklearn/utils/validation.py:
```diff
@@ -2148,9 +2155,10 @@ def _check_sample_weight(
               then the dtype is preserved.
-            - Else the output has NumPy's default dtype: `np.float64`.
+            - Otherwise, the output has the highest precision floating point dtype
+              supported by the array namespace/device of the input arrays.
```
```suggestion
              supported by the array namespace and device of the input arrays.
```

lithomas1 committed:
Merge branch 'main' of github.com:scikit-learn/scikit-learn into roc-curve

lithomas1 committed:
add tests and address comments

lithomas1 commented:
Added some more tests. I think this should be ready for another pass now.
Thanks for the reviews!

OmarManzoor reviewed (approved):
Otherwise looks good. Thanks @lithomas1

lucyleeow commented on sklearn/utils/extmath.py:
```diff
@@ -1232,10 +1241,30 @@ def stable_cumsum(arr, axis=None, rtol=1e-05, atol=1e-08):
+        # np.cumsum calculates cumsum across flattened array
+        # for axis=None, which is why we flatten here.
+        # (if we don't flatten, cumulative_sum below will raise error)
+        arr = xp.reshape(arr, (-1,))
```
What about adding:
```suggestion
        arr = xp.reshape(arr, (-1,))
        axis=0
```
Avoids the `if axis is None:` below

lithomas1 replied:
Thanks, that's much nicer than what I had.

lucyleeow commented on sklearn/utils/extmath.py:
```diff
@@ -1232,10 +1241,30 @@ def stable_cumsum(arr, axis=None, rtol=1e-05, atol=1e-08):
+    out = xp.cumulative_sum(arr, axis=axis, dtype=max_float_dtype)
+    expected = xp.sum(arr, axis=axis, dtype=max_float_dtype)
+    # workaround to get last element
+    # PyTorch doesn't support this ATM
+    # https://github.com/pytorch/pytorch/issues/146211
```
Maybe we could add "TODO" here so we remember to come back to it once supported?

lithomas1 replied:
Updated.

lucyleeow reviewed (commented):
Thanks @lithomas1 ! This looks good, I just had some nits and a comment about tests.

Question for my education, in general, should we be using `_max_precision_float_dtype` instead of `_find_matching_floating_dtype` for metrics...?

lithomas1 committed:
address feedback

lithomas1 commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -1592,48 +1596,96 @@ def test_check_psd_eigenvalues_invalid(lambdas, err_type, err_msg):
+@pytest.mark.parametrize("y_true", [[0], [0, 1], [-1, 1], [1, 1, 1], [-1, -1, -1]])
+def test_check_pos_label_consistency(array_namespace, device, dtype, y_true):
+    xp = _array_api_for_tests(array_namespace, device)
+    with config_context(array_api_dispatch=True):
```
Ended up splitting the test as Lucy recommended.

lithomas1 commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -1592,48 +1596,96 @@ def test_check_psd_eigenvalues_invalid(lambdas, err_type, err_msg):
+    "array_namespace,device,dtype", yield_namespace_device_dtype_combinations()
+)
+@pytest.mark.parametrize("array_api_dispatch", [True, False])
+def test_check_sample_weight(array_namespace, device, dtype, array_api_dispatch):
```
I ended up copy-pasting the test to make a non array API version. Let me know if you had in mind something else.

betatim replied:
We try to have `array_api` in tests that should run when you want to test "array API stuff" so that you can select them with `pytest -k array_api`. The "CUDA CI" job does this as well. So I think we should try and keep doing this (aka do what you did here).

Maybe one day it becomes unwieldy or annoying, but I think for now it is super good enough.

lithomas1 commented:
Sorry for the slow reply, I switched to using a different laptop for open source stuff.

> Thanks @lithomas1 ! This looks good, I just had some nits and a comment about tests.
> 
> Question for my education, in general, should we be using `_max_precision_float_dtype` instead of `_find_matching_floating_dtype` for metrics...?

I think _max_precision_float_dtype is used when creating an singular array. It exists since libraries like PyTorch (with MPS) sometimes don't support float64. I've never used _find_matching_floating_dtype - it seems like it would be used so you can upcast inputs to the highest precision of the input arrays to a function. 

Maybe @OmarManzoor would know more. I think he recommended `_max_precision_float_dtype` to me initially.

lucyleeow reviewed (commented):
Thanks, just some nits!

lithomas1 committed:
Address comments

lithomas1 committed:
Merge branch 'main' into roc-curve

lithomas1 committed:
Missing device call

lucyleeow reviewed (approved):
Thank you for your persistence @lithomas1 ! I will merge in a few days just in case @OmarManzoor wants to give it a final look.

lucyleeow commented:
@lithomas1 looks like we have some failures on the CUDA environment

lithomas1 committed:
address comments

lithomas1 committed:
Merge branch 'main' into roc-curve

lithomas1 committed:
fix CI

lucyleeow commented on sklearn/utils/extmath.py:
```diff
@@ -1232,10 +1241,29 @@ def stable_cumsum(arr, axis=None, rtol=1e-05, atol=1e-08):
+    # TODO: remove once issue is fixed
+    # https://github.com/pytorch/pytorch/issues/146211
+    last_elem_idx = xp.asarray(out.shape[axis] - 1)
+    if not xp.all(
+        xpx.isclose(
+            xp.take(out, xp.asarray([last_elem_idx], device=device), axis=axis),
+            expected,
+            rtol=rtol,
+            atol=atol,
+            equal_nan=True,
+        )
     ):
```
@lithomas1 seems that this part is not working (CI failure shows no warning emitted for cupy)

lucyleeow replied:
Am at a lose as to what is causing this failure. The docs do talk about automatic [type promotion](https://docs.cupy.dev/en/stable/user_guide/difference.html#type-promotion):

> CuPy automatically promotes dtypes of [cupy.ndarray](https://docs.cupy.dev/en/v4.5.0/reference/generated/cupy.ndarray.html#cupy.ndarray) s in a function with two or more operands

but the failure occurs even when `dtype` is float64.

Maybe others can shed some light?

OmarManzoor replied:
It still failed in one case. @betatim @lesteve What do you think about this? This warning is not triggered easily.

ogrisel replied:
I think I would be in favor of not calling scikit-learn's `stable_cumsum` at all in functions with array API support, and instead directly call `xp.cumsum`.

Once all the code base has been migrated to remove all the calls to `stable_cumsum` we can just deprecate this function. The extra call to `np/xp.sum` is adding some unnecessary overhead for no value for the end user: if the `cumsum` result happens to be unstable, the users cannot do anything about it, so I don't really see the point of this warning.

OmarManzoor replied:
In order to merge this PR should we remove the warning code altogether or should we keep the code but stop testing for it?

lucyleeow commented on sklearn/utils/tests/test_extmath.py:
```diff
@@ -1119,3 +1119,36 @@ def test_randomized_range_finder_array_api_compliance(array_namespace, device, d
+@pytest.mark.parametrize("axis", [0, 1, None])
+def test_stable_cumsum_array_api(array_namespace, device, dtype, axis):
+    xp = _array_api_for_tests(array_namespace, device)
+    arr_np = np.asarray(
+        [[1, 2e-9, 3e-9] * int(1e6)],
+    )
+    arr_xp = xp.asarray(
+        arr_np, dtype=getattr(xp, dtype) if dtype is not None else dtype
+    )
```
Why not do:

```suggestion
    arr_np = np.asarray(
        [[1, 2e-9, 3e-9] * int(1e6)], dypte=dtype
    )
    arr_xp = xp.asarray(arr_np)
```
?

OmarManzoor commented on sklearn/utils/tests/test_extmath.py:
```diff
@@ -1119,3 +1119,36 @@ def test_randomized_range_finder_array_api_compliance(array_namespace, device, d
+            np.cumsum(arr_np, axis=axis),
+        )
+
+    arr = xp.asarray(
+        [np.random.RandomState(0).rand(1000000)], dtype=xp.float32, device=device
+    )
```
```suggestion
    arr = xp.asarray([np.full(1000, 100.5)], dtype=xp.float64, device=device)
    arr[-1, -1] += 0.01
```

OmarManzoor committed:
Update sklearn/utils/tests/test_extmath.py

OmarManzoor commented:
> For the particular case of this PR, I think we should revert the changes under `sklearn/utils` and just call `xp.cumsum` directly in the code that computes the ROC curve.

That sounds good. Thanks.

ogrisel commented:
As shown by @lesteve in https://github.com/scikit-learn/scikit-learn/issues/31533#issuecomment-2967062437, casting to `xp.float64` before computing the cumsum can be required to get `roc_auc_score` to return the correct value on some dataset. This PR should add a non-regression test based on Loïc's comment if it's not already present in our test suite.

OmarManzoor committed:
Merge branch 'main' into roc-curve

OmarManzoor committed:
Use xp.cumulative_sum and add a non regression test for roc_curve precision with float32 weights

OmarManzoor commented:
@lesteve Do you have any suggestions on how we can keep this test?

ogrisel reviewed (approved):
Given the cost to maintain a proper non-regression test, I think an inline comment will do.

OmarManzoor committed:
Update sklearn/metrics/_ranking.py

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

OmarManzoor committed:
Update sklearn/metrics/tests/test_common.py

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

OmarManzoor committed:
Remove slow test

OmarManzoor committed:
Merge branch 'main' into roc-curve
